### PR TITLE
[Numpy] Random.gamma() implemented

### PR DESCRIPTION
**Pull request by xidulu**
## Description ##
As title, implementation detail is described in https://github.com/apache/incubator-mxnet/issues/15928

`mx.np.random.gamma()` outperforms `ndarray.random.gamma()` on GPU. Also, it supports broadcastable parameters. However, there exists some performance issues on **CPU**, more specifically, slower than native Numpy.
I'm currently investigating into a proper implementation suitable for CPU backend. 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Feature1, tests, (and when applicable, API doc)
- [x] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/ndarray/numpy/random.py [hunk 1]
```diff
@@ -319,6 +319,63 @@ def choice(a, size=None, replace=True, p=None, ctx=None, out=None):
...
+        Drawn samples from the parameterized gamma distribution.
+
+    The Gamma distribution is often used to model the times to failure of
+    electronic components, and arises naturally in processes for which the
+    waiting times between Poisson distributed events are relevant.
+    """
+    from ...numpy import ndarray as np_ndarray
+    input_type = (isinstance(shape, np_ndarray), isinstance(scale, np_ndarray))
+    if dtype is None:
+        dtype = 'float32'
```
Comment (haojin2) [inline python/mxnet/ndarray/numpy/random.py]: `dtype = np.float32`

File: src/operator/numpy/random/np_gamma_op.cc [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+#include "./np_gamma_op.h"
+
+namespace mxnet {
+namespace op {
+
+DMLC_REGISTER_PARAMETER(NumpyGammaParam);
+
+inline bool NumpyGammaOpType(const nnvm::NodeAttrs &attrs,
+                                   std::vector<int> *in_attrs,
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.cc]: alignment.

File: src/operator/numpy/random/np_gamma_op.h [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+  DMLC_DECLARE_PARAMETER(NumpyGammaParam) {
+    DMLC_DECLARE_FIELD(shape);
+    DMLC_DECLARE_FIELD(scale);
+    DMLC_DECLARE_FIELD(size)
+        .set_default(dmlc::optional<mxnet::Tuple<int>>())
+        .describe("Output shape. If the given shape is, "
+                  "e.g., (m, n, k), then m * n * k samples are drawn. "
+                  "Default is None, in which case a single value is returned.");
+    DMLC_DECLARE_FIELD(ctx)
+    .set_default("xpu")
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.h]: default should be `cpu`

File: src/operator/numpy/random/np_gamma_op.h [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+  dmlc::optional<float> shape;
+  dmlc::optional<float> scale;
+  std::string ctx;
+  int dtype;
+  dmlc::optional<mxnet::Tuple<int>> size;
+  DMLC_DECLARE_PARAMETER(NumpyGammaParam) {
+    DMLC_DECLARE_FIELD(shape);
+    DMLC_DECLARE_FIELD(scale);
+    DMLC_DECLARE_FIELD(size)
+        .set_default(dmlc::optional<mxnet::Tuple<int>>())
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.h]: 2-space indentation.

File: src/operator/numpy/random/np_gamma_op.h (hunk 3 above)
Comment (haojin2) [reply src/operator/numpy/random/np_gamma_op.h]: and, 2-space indentation

File: src/operator/numpy/random/np_gamma_op.h [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+        .describe("Output shape. If the given shape is, "
+                  "e.g., (m, n, k), then m * n * k samples are drawn. "
+                  "Default is None, in which case a single value is returned.");
+    DMLC_DECLARE_FIELD(ctx)
+    .set_default("xpu")
+    .describe("Context of output, in format [xpu|xpu|xpu_pinned](n)."
+              " Only used for imperative calls.");
+    DMLC_DECLARE_FIELD(dtype)
+    .add_enum("float32", mshadow::kFloat32)
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.h]: 2-space indentation.

File: src/operator/numpy/random/np_gamma_op.h [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+
+template <int ndim, typename IType, typename OType, typename FType>
+struct gamma_kernel {
+  MSHADOW_XINLINE static void Map(index_t i,
+                                  const Shape <ndim> &lstride, const Shape <ndim> &hstride,
+                                  const Shape <ndim> &oshape,
+                                  IType *shape, IType *scale,
+                                  FType *uniforms, FType *normals,
+                                  OType *out) {
+  Shape<ndim> coord = unravel(i, oshape);
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.h]: 2-space indentation for the whole function body.

File: src/operator/numpy/random/np_gamma_op.h [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+  auto lidx = static_cast<index_t>(dot(coord, lstride));
+  auto hidx = static_cast<index_t>(dot(coord, hstride));
+  IType shape_value = shape[lidx];
+  IType scale_value = scale[hidx];
+  // map phase
+  GammaTransform<IType, FType>(shape_value, scale_value,
+                                uniforms + i * M, normals + i * M);
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.h]: alignment

File: src/operator/numpy/random/np_gamma_op.h [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+  flag[0] = 1;
+  Shape<ndim> coord = unravel(i, oshape);
+  auto lidx = static_cast<index_t>(dot(coord, lstride));
+  auto hidx = static_cast<index_t>(dot(coord, hstride));
+  IType shape_value = shape[lidx];
+  IType scale_value = scale[hidx];
+  if (out[i] < 0) {
+    // map phase
+    GammaTransform<IType, FType>(shape_value, scale_value,
+                                  uniforms + i * M, normals + i * M);
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.h]: alignment

File: src/operator/numpy/random/np_gamma_op.h [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+
+template<typename OType, typename FType>
+struct CheckSuccessKernel {
+  MSHADOW_XINLINE static void Map(int i, OType* out, FType* flag) {
+    if (out[i] < 0) {
+      flag[0] = -1.0;
+    }
+  }
+};
+
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.h]: can we de-dup the kernels below?
seems like the kernels without `_r` only lacks the extra `flag[0] = 1` operation, so you can probably only have the ones with `_r` and have the `flag` pointer default to `nullptr`, and wrap the `flag[0] = 1` with:
```c++
if (flag != nullptr) {
  flag[0] = 1;
}
```

File: src/operator/numpy/random/np_gamma_op.h [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+        (OType)GammaReduce<float, FType>(shape_value, uniforms);
+      out[i] = sample;
+      }
+    }
+};
+}  // namespace mxnet_op
+
+template <typename xpu, typename FType>
+void NumpyGammaForward(const nnvm::NodeAttrs &attrs, const OpContext &ctx,
+                         const std::vector<TBlob> &inputs,
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.h]: alignment

File: src/operator/numpy/random/np_gamma_op.h [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+    // int type_flag = param.t;
+    if (param.shape.has_value()) {
+      scalar_pos = 0;
+      scalar_value = param.shape.value();
+    } else {
+      scalar_pos = 1;
+      scalar_value = param.scale.value();
+    }
+    MSHADOW_TYPE_SWITCH(inputs[0].type_flag_, IType, {
+      MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, OType, {
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.h]: Seems like your output type can only be floating-point numbers, so you can use `MSHADOW_REAL_TYPE_SWITCH` instead.

File: src/operator/numpy/random/np_gamma_op.h [hunk 12]
```diff
@@ -0,0 +1,422 @@
...
+            s, outputs[0].Size(), lstride, hstride, oshape,
+            inputs[0].dptr<IType>(), inputs[1].dptr<IType>(),
+            uniform_tensor.dptr_, normal_tensor.dptr_,
+            outputs[0].dptr<OType>(), failure_indicator_device);
+          failure_indicator = 1.0;
+          Kernel<CheckSuccessKernel<OType, FType>, xpu>::Launch(
+              s, outputs[0].Size(), outputs[0].dptr<OType>(),
+              failure_indicator_device);
+          _copy<xpu>(s, &failure_indicator, failure_indicator_device);
+          }
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.h]: I think your code could be simplified like this:
```c++
FType failure_indicator = 1.0;
bool flag = false;
do {
  if (flag) {
    prnd->SampleUniform(&uniform_tensor, 0, 1);
    prnd->SampleGaussian(&normal_tensor, 0, 1);
  }

  mxnet_op::Kernel<gamma_kernel_r<NDim, IType, OType, FType>, xpu>::Launch(
    s, outputs[0].Size(), lstride, hstride, oshape,
    inputs[0].dptr<IType>(), inputs[1].dptr<IType>(),
    uniform_tensor.dptr_, normal_tensor.dptr_,
    outputs[0].dptr<OType>(), flag ? failure_indicator_device : nullptr);

  failure_indicator = 1.0;

  Kernel<CheckSuccessKernel<OType, FType>, xpu>::Launch(
    s, outputs[0].Size(), outputs[0].dptr<OType>(),
    failure_indicator_device);

  _copy<xpu>(s, &failure_indicator, failure_indicator_device);
  flag = true;
} while (failure_indicator < 0)
```

Comment (haojin2) [reply src/operator/numpy/random/np_gamma_op.h]: assuming you have merged the kernels with and without `_r` suffix

File: src/operator/numpy/random/np_gamma_op.h [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+          mshadow::Shape<NDim> stride =
+              mxnet_op::calc_stride(new_lshape.get<NDim>());
+          bool in_resample_stage = false;
+          do {
+            if (in_resample_stage) {
+              prnd->SampleUniform(&uniform_tensor, 0, 1);
+              prnd->SampleGaussian(&normal_tensor, 0, 1);
+            }
+            mxnet_op::Kernel<gamma_one_scalar_kernel<NDim, IType, OType, FType>, xpu>::Launch(
+                             s, outputs[0].Size(), scalar_pos, stride, oshape,
```
Comment (haojin2) [inline src/operator/numpy/random/np_gamma_op.h]: indentation.
